File: src/com/texnedo/MergeIntervals.java
package com.texnedo;


import com.texnedo.utils.Interval;

import java.util.*;

class MergeIntervals {

    static class IntervalComparator implements Comparator<Interval> {
        @Override
        public int compare(Interval l, Interval r) {
            return Integer.compare(l.start, r.start);
        }
    }

    public static void main(String[] arr) {
        List<Interval> list = new ArrayList<>();
        list.add(new Interval(4, 5));
        list.add(new Interval(1, 4));
        System.out.println(merge(list));
    }

    public static List<Interval> merge(List<Interval> intervals) {
        if (intervals == null || intervals.size() <= 1) {
            return intervals;
        }
        intervals.sort(new IntervalComparator());
        ListIterator<Interval> iterator = intervals.listIterator();
        LinkedList<Interval> result = new LinkedList<>();
        while(iterator.hasNext()) {
            if (result.size() == 0) {
                result.addLast(iterator.next());
            } else {
                Interval current = iterator.next();
                Interval last = result.peekLast();
                if (current.start > last.end) {
                    result.addLast(current);
                    continue;
                }
                if (current.end <= last.end) {
                    continue;
                }
                last.end = current.end;
            }
        }
        return result;
    }
}

File: algo-python/MergeIntervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        points = []
        for item in intervals:
            points.append((item[0], -1))
            points.append((item[1], 1))
        points.sort(key=lambda x: (x[0], x[1]))
        stack = []
        result = []
        for current in points:
            if len(stack) == 0:
                stack.append(current)
            else:
                last = stack[-1]
                if last[1] == -1 and current[1] == -1:
                    stack.append(current)
                elif last[1] == -1 and current[1] == 1:
                    last = stack.pop()
                    if len(stack) == 0:
                        result.append([last[0], current[0]])
                else:
                    raise Exception()
        return result


if __name__ == '__main__':
    s = Solution()
    print(s.merge([[1,3],[2,6],[8,10],[15,18]]))
    print(s.merge([[1,4],[4,5]]))